### ISSUE TITLE
Add Fast-Forward to PEFT

### ISSUE THREAD

**Issue by AdirRahamim** [huggingface/peft#2683]
### Feature request

Hello, 

We propose to add fast forward to peft. Fast forward is a novel method to accelerate low-rank training recently published in Fast Forwarding Low-Rank Training (https://aclanthology.org/2024.emnlp-main.535.pdf).
Fast forward accelerates training by alternating between fast-forward phases and regular optimizer steps. In fast forward phase, we keep updating the model with the last update step multiple times. How many? until the loss stops improving on a very tiny validation set. Upon that, we continue with a few regular optimizer steps.
In the original paper, the authors evaluated fast forward on different model sizes (from 1.4B to 8B models) and on different tasks (medical tuning, instruction tuning and chat tuning) and showed that fast forward can save between 41\%-84\% of training FLOPS (depending on the task and model). Similar trends are for training time savings and for DoRA.

**Key Benefits**

* Saves a large amount of resources used during training.
* Does not harm model performance.
* Works on various model sizes and tasks.
* Works with LoRA and extensions such as DoRA.

### Your contribution

Open source implementation of Fast forward: https://github.com/AdirRahamim/Fast-Forward.
I'm available to prepare a PR for PEFT.

**Comment by BenjaminBossan** [comment]
@githubnemo who has already tested this with PEFT.

**Comment by githubnemo** [comment]
Hey :) 

This is a neat coincidence, I recently implemented fast-forward training in our MetaMathQA benchmark to see how it fares. I just pushed the changes in [this PR](https://github.com/huggingface/peft/pull/2693) for discussion. The results were so-so and overall there weren't a lot of fast-forward updates. I am curious if I did something fundamentally wrong or if the results are expected.

In any case, I'd be happy to guide you through the PR! I think we have two options, either providing a `fast_forward_training(peft_model, optimizer, step, **fast_forward_kwargs)` function that serves as a training 'callback' to be incorporated in every step (e.g., for `transformers.Trainer` you'd then have to create an `OptimizerCallback` and call this function), or, and that might be easier to integrate with common training frameworks, a wrapper for the optimizer, e.g. `FastForwardOptimizer(model, optimizer, **fast_forward_kwargs)`. I'm open to other suggestions as well, though.

**Comment by AdirRahamim** [comment]
Hi! 
That is indeed a nice coincidence :) There are several things I can think of that might cause it to work less well:
1. You have to perform several optimizer steps between each FF phase; in our experiments 6 were enough but maybe more are needed in your setting.
2. Did you use LoRA training? In our experiments we saw that FF does not work with full FT.

I think the second option is a more user-friendly option and an option that will be more easily integrated into existing codes and easy to change.
I'll be happy to walk around and hear what you did in your PR :)

**Comment by githubnemo** [comment]
> Hi! That is indeed a nice coincidence :) There are several things I can think of that might cause it to work less well:
> 
> 1. You have to perform several optimizer steps between each FF phase; in our experiments 6 were enough but maybe more are needed in your setting.
> 
> 2. Did you use LoRA training? In our experiments we saw that FF does not work with full FT.

Best to look at the PR but I used LoRA training with 6 steps between FF and backprop. Changing the termination logic had an impact but mostly negative (more details in the PR description). Feel free to comment on the PR, maybe we can find something that benefits the implementation of FF in PEFT, maybe we don't :)

> I think the second option is a more user-friendly option and an option that will be more easily integrated into existing codes and easy to change.

Sure. Would you like to create a PR for that?
